---
title: "Assignment 2"
author: "Benjamin Tudor Price | TUDORPR1"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 2
bibliography: refs.bib
csl: https://raw.githubusercontent.com/citation-style-language/styles/master/research-institute-for-nature-and-forest.csl
---

## Format Variants 
```{r}
get_variant_types <- function(org, mut){
  # Returns the class of mutation given the original and mutant variants.
  muts <- unlist(strsplit(mut, split = ','))
  types <- character()
  
  if (all(nchar(muts) == nchar(org))){
    return('substitution')
  }
  
  if (any(nchar(muts) > nchar(org))){
    types <- c(types, 'insertion')
  }
  
  if (any(nchar(muts) < nchar(org))){
    types <- c(types, 'deletion')
  }
  
  if (any(nchar(muts) - nchar(org) %% 3 != 0)){
    types <- c(types, 'frameshift')
  }
  
  return(types)
}

to_igv_coords <- function(chr, pos){
  # Get IGV Coords.
  return(paste0(c(chr, ':', as.character(pos)), collapse='', sep=''))
}


library(readr)
vars_org <- read_csv("mgy21_S101_R1_001_analyzed.csv")
vars <- data.frame(chr=vars_org$`#CHROM`, pos = vars_org$POS, org=vars_org$REF,
                   mut=vars_org$ALT, qual=vars_org$QUAL)

vars$coords <- c()
for (i in seq_along(vars$chr)){
  vars$coords[i] <- to_igv_coords(vars$chr[i], vars$pos[i])
}

muts <- list()

for (i in seq_along(vars$chr)){
  muts[i] <-  list(get_variant_types(vars$org[i], vars$mut[i]))
}

is_of_mut_type <- function(types, mut){
  # Return whether each mut types in <types> is of type <mut>.
  return(unlist(lapply(types, function(x){mut %in% x})))
}

num_insertion <- sum(is_of_mut_type(muts, 'insertion'))
num_substitution <- sum(is_of_mut_type(muts, 'substitution'))
num_deletion <- sum(is_of_mut_type(muts, 'deletion'))
num_pot_framshift <- sum(is_of_mut_type(muts, 'frameshift'))

```


# Get ORFs and Their Positions
```{r}

library(seqinr)
orfs <- read.fasta(file = "orf_trans.fasta",
  seqtype = c("AA"), set.attributes = FALSE, legacy.mode = TRUE, seqonly = FALSE, 
  whole.header = TRUE)

get_row <- function(orf_desc){
  orf_attrs <- unlist(strsplit(orf_desc[[1]], ','))
  gene_info <- unlist(strsplit(orf_attrs[1], ' '))
  
  gene_name <- gene_info[2]
  
  loc_info <- unlist(strsplit(orf_attrs[2], ' '))
  
  chr <- paste(c('chr', loc_info[3]), sep='', collapse = '')
  pos_info <-  as.integer(unlist(strsplit(loc_info[5], '-')))
  start <- min(pos_info)
  stop <- max(pos_info)
  
  return(list(gene_name, chr, start, stop))
}
orf_df <- do.call(rbind.data.frame, lapply(names(orfs), get_row))
names(orf_df) <- c('name', 'chr', 'start', 'stop')
names(orfs) <- c()
orf_df$seq <- lapply(orfs, paste0, sep='', collapse ='')

```


## Find All Variants that Lie in ORFs
```{r}
for (var_ind in seq_along(vars$chr)){
  var_chr <- vars$chr[var_ind]
  var_pos <- vars$pos[var_ind]
  
  matching_orfs <- which(orf_df$chr == var_chr & 
                         orf_df$start <= var_pos &
                         var_pos <= orf_df$stop)
  orf_names <- orf_df$name[matching_orfs]
  orf_names <- paste0(orf_names, collapse = ', ')
  vars$orfs[var_ind] <- orf_names
}

mut_counts <- table(vars$orfs)
mut_counts <- mut_counts[order(mut_counts, decreasing = TRUE)][2:length(mut_counts)]


library(knitr)
kable(mut_counts, col.names = c('Gene Name', '# of Variants'))

```


## Sequences of Notable ORFs
```{r}
num_names <- 4

notable_gene_names <- names(mut_counts)[1:(num_names + 1)]
sel <- match(notable_gene_names, orf_df$name)
seqs <- orf_df$seq[sel]

notable_genes <- data.frame(name=notable_gene_names, seq=unlist(seqs))

for (i in seq_along(notable_genes$name)){
  ind <- sel[i]
  notable_genes$loc[i] <-  to_igv_coords(orf_df$chr[ind], orf_df$start[ind])
}

```


## Analysis of all Replicates Runs
```{r}
csvs <- list.files('all_variant_calls')
variant_call_dfs <- list()

# Parse all variant calls.
for (i in seq_along(csvs)){
  csv_name <- csvs[i]
  csv_path <- paste('all_variant_calls/', csv_name, collapse = '', sep='')
  print(csv_path)
  csv_org <- read_csv(csv_path)
  csv_formatted <- data.frame(chr=csv_org$`#CHROM`, pos = csv_org$POS, org=csv_org$REF,
                   mut=csv_org$ALT)
  variant_call_dfs[i] <- list(csv_formatted)
}

install.packages("dplyr")
library(dplyr)

# Get all unique mutations.
all_muts <- do.call(rbind.data.frame, variant_call_dfs)
unique_muts <- distinct(all_muts)

# Count copies across datasets.
copies <- numeric(nrow(unique_muts))
for (i in seq(1, nrow(unique_muts))){
  mut <- trimws(unique_muts[i,])
  
  copies[i] <- sum(apply(all_muts, MARGIN=1, FUN = function(x){all(trimws(x)==mut)}))
  print(i)
  print(copies[i])
}

# Order by position, then number of counts.
unique_muts$copies <- copies
unique_muts <- unique_muts[order(unique_muts$pos),]
unique_muts <- unique_muts[order(unique_muts$chr),]

# Find mutations that map to specific orfs.
for (var_ind in seq_along(unique_muts$chr)){
  var_chr <- unique_muts$chr[var_ind]
  var_pos <- unique_muts$pos[var_ind]
  
  matching_orfs <- which(orf_df$chr == var_chr & 
                         orf_df$start <= var_pos &
                         var_pos <= orf_df$stop)
  orf_names <- orf_df$name[matching_orfs]
  orf_names <- paste0(orf_names, collapse = ', ')
  unique_muts$orfs[var_ind] <- orf_names
}

orf_to_counts <- c()
orf_names <- unique(unique_muts$orfs)

for (i in seq_along(orf_names)){
  orf <- orf_names[i]
  sel <- unique_muts$orfs == orf
  orf_to_counts[i] <- sum(unique_muts$copies[sel])
  names(orf_to_counts)[i] <- orf
}
names(orf_to_counts)[which(names(orf_to_counts) == '')] <- 'Not in an ORF'
orf_to_counts <- orf_to_counts[order(orf_to_counts, decreasing = TRUE)]

hist(unique_muts$copies, main='Histogram of Copy Number', xlab = 'Number of Copies Across all Students', ylab="Variants")
write_csv(unique_muts, 'Unique Variants.csv')


```


## Filter Likely G0 Variants
```{r}
orf_names <- names(orf_to_counts)
orf_stats <- data.frame(name=character(), 
                        num_dups=numeric(), 
                        num_unique=numeric())

dup_threshold <- 10

for (name in orf_names){
  sel <- unique_muts$orfs == name
  if (name == 'Not in an ORF'){
    sel <- unique_muts$orfs == ""
  }
  dups <- unique_muts$copies[sel] >= dup_threshold
  uniques <- unique_muts$copies[sel] < dup_threshold
  orf_stats <- rbind(orf_stats, list(name, sum(dups), sum(uniques)))
}

names(orf_stats) <- c('name', 'num_dups', 'num_unique')

# Filter genes out with more than 4 duplicates.
orf_stats <- orf_stats[orf_stats$num_dups == 0,]
orf_stats <- orf_stats[order(orf_stats$num_unique, decreasing = TRUE),]
write_csv(orf_stats, 'Promising ORFs.csv')
```
FLO9, 	32
FLO1, 	10
YHL048C, -A	9
PKC1, 	5
CTR9, 	3
HSM3, 	3
MTM1, 	3
YBL100W, -A, YBL100W-B	3
YGK3, 	3
YHL050C, 	3
CSS1, 	2
GAL2, 	2
HFA1, 	2
NUM1, 	2
PYC1, 	2
SSA1, 	2
YCR024C, -B	2
APC1, 	1
ASG1, 	1
BUD5, 	1
CAN1, 	1
DOP1, 	1
DPB2, 	1
EBS1, 	1
ESC1, 	1
GPB2, 	1
GPI1, 	1
HIS3, 	1
KEL1, 	1
KRE33, 	1
KRE5, 	1
MDL1, 	1
MEH1, 	1
MKK1, 	1
MNN1, 	1
NAB6, 	1
ORM2, 	1
PFK2, 	1
PKH1, 	1
PMP1, 	1
POS5, 	1
PPX1, 	1
RDH54, 	1
RRG8, 	1
RSC30, 	1
RTP1, 	1
SAC7, 	1
SAL1, 	1
SCH9, 	1
SEC1, 	1
SHC1, 	1
SRD1, 	1
SST2, 	1
TAX4, 	1
TOH1, 	1
TRP4, 	1
VID28, 	1
YDR210C, -D	1
YFL066C, 	1
YHL041W, 	1
YLL067C, 	1
YLR157W, -D	1
YLR302C, 	1
YMR027W, 	1
YMR317W, 	1
YNL054W, -A, YNL054W-B	1, 